### PR TITLE
Change shutdown approach

### PR DESCRIPTION
**Pull request by stevehodgkiss**
There was an issue rescheduling jobs in `the_end_is_nigh`.

Instead of manually rescheduling jobs, let's rely on a pipeline's retry config. The changes here mean that on shutdown, we only tell the agent to stop accepting new work with `agents.stop`.

### CODE REVIEW

File: spec/buildkite_agent_spec.rb [hunk 1]
```diff
@@ -2,11 +2,21 @@
 
 describe SpotBuild::BuildkiteAgent do
   let(:org_slug) { "envato" }
+  let(:pipeline) { "my-app" }
   subject { described_class.new('deadbeef', org_slug) }
   
   let(:last_response_stub) { instance_double(Sawyer::Response) }
-  let(:buildkit_stub) { double("Buildkit", :agents => agent_stubs) }
+  let(:buildkit_stub) { instance_double("Buildkit::Client", :agents => agent_stubs) }
```
Comment (patrobinson) [inline spec/buildkite_agent_spec.rb:8]: 💯

**Review by patrobinson — APPROVED** [review]

File: lib/spot_build/buildkite_agent.rb [hunk 2]
```diff
@@ -9,29 +9,46 @@ def initialize(token, org_slug)
...
+        reschedule_job(agent.job)
+      end
+      agents.count
     end
 
-    def stop(force="false")
-      return unless agent_running?
-      @client.stop_agent(@org_slug, agent_id, "{\"force\": #{force}}")
+    def stop_agent(agent, force: false)
+      @client.stop_agent(@org_slug, agent.id, "{\"force\": #{force}}")
```
Comment (ali-graham) [inline lib/spot_build/buildkite_agent.rb:24]: That hash shouldn't be stringified:

```ruby
def stop_agent(org, id, options = {})
```

https://github.com/Shopify/buildkit/blob/master/lib/buildkit/client/agents.rb#L36

Comment (patrobinson) [reply lib/spot_build/buildkite_agent.rb:24]: Pretty sure I tried to use it without stringifying it and it didn't work.

File: lib/spot_build/buildkite_agent.rb [hunk 3]
```diff
@@ -9,29 +9,46 @@ def initialize(token, org_slug)
...
+
     def reschedule_job(job)
       return if job.nil?
-      @client.retry_job(@org_slug, job_pipeline(job[:build_url]), job_build(job[:build_url]), job[:id])
+      retry_error(Buildkit::BadRequest, RETRY_MESSAGE) do
+        @client.retry_job(@org_slug, job_pipeline(job[:build_url]), job_build(job[:build_url]), job[:id])
+      end
+    end
+
+    def retry_error(error_class, message_regex, sleep: 2, retries: 20)
```
Comment (patrobinson) [inline lib/spot_build/buildkite_agent.rb]: It seems like this sleep variable is not being used?

**Review by patrobinson — APPROVED** [review]

File: lib/spot_build/buildkite_agent.rb (hunk 2 above)
Comment (stevehodgkiss) [reply lib/spot_build/buildkite_agent.rb:24]: Yeah a string works for `data` in this call (although the variable is called `options` initially`) https://github.com/lostisland/sawyer/blob/master/lib/sawyer/agent.rb#L96

**Review by patrobinson — COMMENTED** [review]
Can we just disable `the_end_is_night` behind a command line flag?

That way we can turn it on/off as needed

**Pull request merged**